Paul Hands Legal Agreements - Added skeleton framework for GMSLA, GMRA and ISDA Master Agreements into Agreement Namespace

regnosys-prod-user opened the pull request.

regnosys-prod-user commented:
# *Legal Agreements - Added skeleton framework for GMSLA, GMRA and ISDA Master Agreements into Agreement Namespace*

_Background_

While the CDM is leading the path in making a universal language to discuss legal agreements, a huge hurdle to get over is that the legal agreements themselves are written inherently differently. While there is overlap, there is NOT enough overlap to allow for e.g. an ISDA master agreement to be represented easily and readily into the system.

This is something the collateral working groups are moving towards with the ISDA foundations model and I believe a similar, heavily typed approach would work for the CDM for an ISDA master agreement, as well as a GMRA and GMSLA

_What is being released?_

As per issue #3206 on Github, we are implementing step one in this contribution. This entails adding the MasterAgreementBase to the master namespace type which is then extended for GMRA, GMSLA and IsdaMasterAgreement respectively. These are all currently empty and will be populated provisionally and initially into the next development release.

MasterAgreementSchedule is remaining as is for the time being to ensure backwards compatability with production, but will be deprecated and changed in a future release PR.

_Review directions_

Changes can be reviewed in PR: [#3624](https://github.com/finos/common-domain-model/pull/3624)

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

lolabeis commented on rosetta-source/src/main/rosetta/legaldocumentation-master-gmra-type.rosetta:
```diff
@@ -0,0 +1,6 @@
+
+import cdm.legaldocumentation.master.*
+
+type Gmra extends MasterAgreementBase : <"The set of elections that define a GMRA">
```
For naming consistency, consider renaming type as: "MasterRepurchaseAgreement"

chrisisla replied:
To GMRA there is also an MRA - I'll leave @tomhealey-icma to comment on what the best namespace/type name should be.

lolabeis commented on rosetta-source/src/main/rosetta/legaldocumentation-contract-type.rosetta:
```diff
@@ -12,9 +15,17 @@ type Agreement: <"Specification of the standard set of terms that define a legal
+type MasterAgreementElections: <"A representation of a master agreement extended from a base agreement of shared elections">
+    gmsla Gmsla (0..1) <"The set of elections that define a GMSLA">
+    gmra Gmra (0..1) <"The set of elections that define a GMRA">
+    isdaMasterAgreement IsdaMasterAgreement (0..1) <"The set of elections that define an ISDA Master Agreement">
```
Rename attribute as simply "isda", for naming consistency.

lolabeis reviewed (commented):
Thanks @paulhands for the contribution. Please see a few detailed review comments inline.

The build is failing, which may be because model snippets in the user doc are no longer in sync with the updated model (this is a built-in safeguard to ensure that documentation is kept up to date when we modify the model). @Oscar-Twomey can help you coordinate with our technical maintenance team here, to identify the source of the build fail and the snippets to update, if that's indeed the reason.

chrisisla commented on rosetta-source/src/main/rosetta/legaldocumentation-master-gmsla-type.rosetta:
```diff
@@ -0,0 +1,6 @@
+
+import cdm.legaldocumentation.master.*
+
+type Gmsla extends MasterAgreementBase : <"The set of elections that define a GMSLA">
```
The MSLA and the GMSLA are two separate documents so I think this would need to be expanded to "GlobalMasterSecuritiesLendingAgreement" in order to prevent confusion.

Also worth noting that I think SIFMA also have a GMSLA so we may need to change the namespace to "isla" in order to identify the types held within as for the ISLA GMSLA.

lolabeis commented on rosetta-source/src/main/rosetta/legaldocumentation-contract-type.rosetta:
```diff
@@ -4,6 +4,9 @@ version "${project.version}"
 import cdm.legaldocumentation.transaction.*
+import cdm.legaldocumentation.master.gmsla.*
+import cdm.legaldocumentation.master.gmra.*
+import cdm.legaldocumentation.master.isdamasteragreement.*
```
In light of @chrisisla 's comment that there could be more than one GMSLA, consider the following:
- use import "alias" to attribute an explicit name to the imported namespace
- use the alias in the path to invoke a type from that namespace
```
import cdm.legaldocumentation.master.isla as isla
...
type MasterAgreementElections:
    islaGmsla isla.GlobalMasterSecuritiesLendingAgreement (0..1)
```
This allows to have another type called `GlobalMasterSecuritiesLendingAgreement` in a different namespace, and the namespace path disambiguates which is which.

I recommend applying the same aliasing for isda and icma as well, just for consistency.

chrisisla replied:
Thanks @lolabeis :)

To make sure I am reading this correctly then I believe that to disambiguate the ISLA GMSLA from the SIFMA one (and others) we should actually create the new namespaces as follows:

`cdm.legaldocumentation.master.isla`
`cdm.legaldocumentation.master.icma`
`cdm.legaldocumentation.master.isda`

So these will hold the GMSLA, GMRA and ISDA Master types respectively.
 
Then, when we import the new namespaces into `legaldocumentation.contract`, we import them as a shortened name and use that as a prefix when we reference them, so:

```
import cdm.legaldocumentation.master.isla as isla
import cdm.legaldocumentation.master.icma as icma
import cdm.legaldocumentation.master.isda as isda
 
type MasterAgreementElections:
  islaGMSLA isla.GlobalMasterSecuritiesLendingAgreement (0..1)
  icmaGMRA icma.GlobalMasterRepoAgreement (0..1)
  isdaMaster isda.MasterAgreement (0..1)
```
We'll need to get this design agreed at the CRWG but I think this looks okay

PayalKhanna commented:
replaced with https://github.com/finos/common-domain-model/pull/3629

lolabeis commented:
Skeleton design agreed at the [CDM Contribution Review Working Group - April 15th, 2025](#3603)

regnosys-prod-user commented:
_What is being released?_

Added new namespaces as agreed in CRWG, added to be deleted comments to redundant namespaces, imported as aliases

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

regnosys-prod-user commented:
_What is being released?_

Internal child renamed from gmra to icmaGmra

> [!NOTE]
> This comment was generated via [Rosetta](https://rosetta-technology.io/).

paulhands commented:
@Oscar-Twomey I have added the comments on namespaces that need deleting because I cannot do that in rosetta.

@PayalKhanna I don't know if more updates will be needed here

Oscar-Twomey commented:
> @Oscar-Twomey I have added the comments on namespaces that need deleting because I cannot do that in rosetta.
> 
> @PayalKhanna I don't know if more updates will be needed here

Thanks Paul, the comments are sufficient.
